go/def.bzl

```diff
@@ -217,19 +217,39 @@ def emit_go_compile_action(ctx, sources, deps, out_lib, extra_objects=[]):
   if _is_external(out_dir):
     out_depth -= 2
   cmds = symlink_tree_commands(out_dir, tree_layout)
+
+  extra_inputs = ctx.files.toolchain
+
+  # cd into the out_dir.
+  cmds.append("export GOROOT=$(pwd)/" + ctx.file.go_tool.dirname + "/..")
+  cmds.append(" ".join(["cd", out_dir]))
+
+  # Apply build tags here.
+  use_build_tags = hasattr(ctx.attr, "filter_tags")
+  cleaned_source_paths = [prefix + _remove_external_prefix(i.path) for i in sources]
+  if use_build_tags:
+    filter_args = [
+        ('../' * out_depth) + ctx.executable.filter_tags.path,
+    ]
+    filter_args.extend(cleaned_source_paths)
+    cmds.append("export GOFILES=\"$({})\"".format(" ".join(filter_args)))
+    extra_inputs += [ctx.executable.filter_tags]
+  else:
+    cmds.append("export GOFILES='{}'".format(" ".join(cleaned_source_paths)))
+
   args = [
-      "cd ", out_dir, "&&",
       ('../' * out_depth) + ctx.file.go_tool.path,
       "tool", "compile",
       "-o", ('../' * out_depth) + out_lib.path, "-pack",
-      "-I", "."
+      "-I", ".",
+      "${GOFILES}",
```

Comment:
I don't think this will correctly handle filenames with spaces. Could you add a TODO here or above? I'll fix this before the next release.